Madness/Parser.swift

```diff
@@ -3,31 +3,31 @@
 /// Convenience for describing the types of parser combinators.
 ///
 /// \param Tree  The type of parse tree generated by the parser.
-public struct Parser<C: CollectionType, Tree> {
+public enum Parser<C: CollectionType, Tree> {
 	/// The type of parser combinators.
 	public typealias Function = (C, C.Index) -> Result
 
 	/// The type produced by parser combinators.
-	public typealias Result = (Tree, C.Index)?
+	public typealias Result = Either<Error<C.Index>, (Tree, C.Index)>
 }
 
 
 /// Parses `input` with `parser`, returning the parse trees or `nil` if nothing could be parsed, or if parsing did not consume the entire input.
-public func parse<C: CollectionType, Tree>(parser: Parser<C, Tree>.Function, input: C) -> Tree? {
-	return parser(input, input.startIndex).map { $1 == input.endIndex ? $0 : nil } ?? nil
+public func parse<C: CollectionType, Tree>(parser: Parser<C, Tree>.Function, input: C) -> Either<Error<C.Index>, Tree> {
+	return parser(input, input.startIndex).flatMap { $1 == input.endIndex ? .right($0) : .left(.leaf("finished parsing before end of input", $1)) }
 }
 
 
 // MARK: - Terminals
 
 /// Returns a parser which never parses its input.
-public func none<C: CollectionType, Tree>() -> Parser<C, Tree>.Function {
-	return const(nil)
+public func none<C: CollectionType, Tree>(string: String = "no way forward") -> Parser<C, Tree>.Function {
```

Comment:
This is a deceptively small change, so take note: it allows grammars to embed information about semantic errors!

That is, you can report better errors by detecting mistakes in the parser itself and failing via `none(…)`:

```swift
let abstraction = %"λ" ++ variable ++ (%"." | none("looks like you forgot a . !"))
```